### ISSUE TITLE
[Meta] View site certificate

### ISSUE THREAD

**Issue by vesta0** [mozilla-mobile/fenix#5073]
### Why/User Benefit/User Problem

As a user, I want to know if a site has been verified and view the certification, so I feel more secure and have more control over what I am browsing.

### Dependencies
- [x] GV exposing [site certificate information](https://bugzilla.mozilla.org/show_bug.cgi?id=1508730)
- [ ] #5892 Error pages has landed

### Acceptance Criteria

(p1) I can tap on the lock icon and open the secure site panel, I see "Verified by: X" underneath "Secure "Connection
(p2) I have access to the site certificate

Fenix UI
![1988930735](https://user-images.githubusercontent.com/42752644/65936915-2b509880-e3ec-11e9-9fe3-522bf44c4900.png)

Fennec UI (includes the site icon - not part of this user story)
![Screenshot_20191001-013101](https://user-images.githubusercontent.com/42752644/65936716-8afa7400-e3eb-11e9-9f7c-8d13f67d6df7.png)

Focus UI (does not have the site icon)
![Screenshot_20191001-012903](https://user-images.githubusercontent.com/42752644/65936723-92ba1880-e3eb-11e9-95ac-15de3b2874cb.png)

**Comment by sblatz** [comment]
@vesta0 is this a must or a should? It's labeled "should" here but "must" in master plan doc.

**Comment by vesta0** [comment]
Hey @sblatz it was a must that I changed to a should after getting feedback from ENG managers. I'll update it.

**Comment by kbrosnan** [comment]
If we get addon support, https://addons.mozilla.org/en-US/firefox/addon/certainly-something/ is an option.

**Comment by sblatz** [comment]
We don't want to depend on the extension because we're still building that out this quarter. However, @kbrosnan said we could port a lot of the code since it was built in-house. 

This has an a-c label because we will need to read metadata about the page, however I think we already have hooks to this to display the SSL information (at least according to @colintheshots).

**Comment by vesta0** [comment]
@lime124 I recommend that we skip UX for this issue and use the Fennec or Fenix UI as reference.

**Comment by april** [comment]
I will note that [Certainly Something](https://github.com/april/certainly-something) already works perfectly on Fennec, and is getting merged into Firefox 71. In theory, it should be relatively straightforward to steal that code.  :)

**Comment by vesta0** [comment]
@AmyYLee will attach a screenshot :)

**Comment by AmyYLee** [comment]
Hi, please reference the line item for Enhanced Tracking Protection in the panel for styling. Thanks!

![Screen Shot 2019-10-03 at 12 58 08 PM](https://user-images.githubusercontent.com/5992746/66147784-85fd1680-e5dd-11e9-98ea-55f5136cbfcf.png)

**Comment by kbrosnan** [comment]
@AmyYLee wrong screenshot? This is for viewing the cert info.

**Comment by vesta0** [comment]
Desktop is about to ship a new in-content certificate viewer (Certainly something add-on) in Firefox 71 and it was developed with mobile in mind. 

It already works on Fenix as it’s shipped with GeckoView. If we add some CSS to make it mobile friendly, then we can just add a link and not do anything.

Need to localize? No

Would A-C package things? Wouldn’t need to as the functionality is part of Gecko.

This is the API / SecurityInformation we’re using in Android Components: https://dxr.mozilla.org/mozilla-central/source/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/GeckoSession.java?q=GeckoSession.java&redirect_type=direct#2786 

https://searchfox.org/mozilla-central/source/mobile/android/modules/geckoview/GeckoViewProgress.jsm#318
 
Next steps:

-File issue for GeckoView to expose the certificate details we need to construct the about:certificate?cert= URL.
-File issue in A-C to make sure this information is exposed and usable in applications / could be the complete link we expose here
-File issue for Desktop/Toolkit to make the existing about:certificate layout responsive and look nice on mobile
 


 

Desktop is about to ship a new in-content certificate viewer (Certainly something add-on) in Firefox 71

Security team developed it with mobile in mind. 

Johann will send me screen shots.

It already works on Fenix as it’s shipped with GeckoView. If we add some CSS to make it mobile friendly, then we can just add a link and not do anything.

Need to localize? No

Would A-C package things? Wouldn’t need to as the functionality is part of Gecko.

This is API / SecurityInformation we’re using in Android Components: https://dxr.mozilla.org/mozilla-central/source/mobile/android/geckoview/src/main/java/org/mozilla/geckoview/GeckoSession.java?q=GeckoSession.java&redirect_type=direct#2786 

https://searchfox.org/mozilla-central/source/mobile/android/modules/geckoview/GeckoViewProgress.jsm#318

Can contributors work on it? Yes

Github issue: https://github.com/mozilla-mobile/fenix/issues/5073
 
Next steps:
File issue for GeckoView to expose the certificate details we need to construct the about:certificate?cert= URL.
File issue in A-C to make sure this information is exposed and useable in applications / could be the complete link we expose here
File issue for Desktop/Toolkit (not sure which component is best) to make the existing about:certificate layout responsive and look nice on mobile

**Comment by sblatz** [comment]
When we add this, let's make sure we add the "view certificate" button in the error page: https://miro.com/app/board/o9J_kwVai6Q=/

**Comment by wisniewskit** [comment]
Note that in [bz1553265](https://bugzilla.mozilla.org/show_bug.cgi?id=1553265), support just landed on error pages to get the failed certificate in the format needed by `about:certificates`. That is, on error pages a link can be formed via:
```
const link = "about:certificate?cert=" +
                document.getFailedCertSecurityInfo().certChainStrings.
                  map(c => encodeURIComponent(c)).
                  join("&cert=")
````

**Comment by vesta0** [comment]
@liuche will investigate whether this is ready to be picked up in the next sprint.

**Comment by liuche** [comment]
This should be ready to work on now that GV site cert bug (see comment 0) has been fixed. WebResponse should include the site certificate information now.

**Comment by liuche** [comment]
@kglazko can coordinate w/ @sblatz when Error pages is landed.

**Comment by liuche** [comment]
@kglazko can you push a WIP PR? We're discussing what the next steps are during sprint planning. Is this actually blocked on Error pages? Looking at this, it seems like the main entry point is from the lock icon in the urlbar.

**Comment by sblatz** [comment]
@vesta0 @liuche I think there's been some confusion on this ticket.

What @kglazko is working on is adding the functionality to click on the lock icon in the toolbar and see who the certificate is "verified by."

I'm not sure we have access to the certificate itself (to be able to show it in the error pages). I will ask AC to comment here if this is available. 

If we're waiting on that, is it critical we have the "view certificate" in error pages? We can bypass them and show the following warning: 

<img width="350" alt="image" src="https://user-images.githubusercontent.com/4400286/73550495-02e1b500-4445-11ea-8ff2-f2ecb225e3ad.png">

**Comment by st3fan** [comment]
I think what we have for the current page is documented at:

https://mozac.org/api/mozilla.components.browser.session/-session/-security-info/

If that is not enough (it only has "Issuer") then please file A-C bugs to expose the data we need to show. (Eg. not sure if "Issuer" is same as "Verified by")

cc @pocmo

**Comment by kglazko** [comment]
I was operating under the assumption that Issuer is the same as Verified By in my PR ^

**Comment by st3fan** [comment]
@wisniewskit is there a way to call `about:certificate` for the valid one of the page currently looked at? I assume that is what we want to display all the details? Instead of writing our own UI?

**Comment by st3fan** [comment]
> I was operating under the assumption that Issuer is the same as Verified By in my PR ^

I think that is probably correct but we need to verify that with the right people because it is an important detail. Maybe @wisniewskit can point us in the right direction?

**Comment by kglazko** [comment]
<img width="404" alt="Screen Shot 2020-01-31 at 8 10 28 AM" src="https://user-images.githubusercontent.com/1676235/73555440-7dd5ac00-4402-11ea-9b31-800bfc2628b9.png">
<img width="401" alt="Screen Shot 2020-01-31 at 8 10 55 AM" src="https://user-images.githubusercontent.com/1676235/73555443-7f06d900-4402-11ea-9097-c8bb28562585.png">
<img width="386" alt="Screen Shot 2020-01-31 at 8 11 11 AM" src="https://user-images.githubusercontent.com/1676235/73555447-80d09c80-4402-11ea-937f-9df71b9ec5e7.png">

**Comment by kglazko** [comment]
Current implementation is in images above. Please let me know if there should be changes.

**Comment by wisniewskit** [comment]
@st3fan, you should be able to get the URL parameter for about:certificate by calling `document.getFailedCertSecurityInfo()` on the SSL error page; I think the sub-object was `failedCertChainStrings` (I recall posting some sample code on the bugzilla bug for this, I'll check when I get back from dinner).

**Comment by wisniewskit** [comment]
@st3fan, ah, silly me, I had actually posted the example code in this very bug in this comment https://github.com/mozilla-mobile/fenix/issues/5073#issuecomment-567539126

**Comment by softvision-miralobontiu** [comment]
Verified as fixed on the latest Nightly build from 2/11 with Google Pixel (Android 10), and Huawei P9 Lite (Android 7). The lock icon, when tapped, displays the "Secure Connection, Verified by..." information.

I've filed #8300 for the visibility of the information displayed when using dark theme.

**Comment by feedback-query122** [comment]
The latest nightly version is showing "verified by:". However, option to view the site certificate is still not available.

**Comment by feedback-query122** [comment]
@softvision-miralobontiu the latest nightly version shows "verified by:". However, option to view the site certificate is not available.

**Comment by softvision-miralobontiu** [comment]
@vesta0 should the user have access to the site certificate?
In Acceptance criteria there is "(p2) I have access to the site certificate" but according to https://github.com/mozilla-mobile/fenix/issues/5073#issuecomment-554108976 issues should be filed in GeckoView and A-C regarding this exposure.
Should I reopen this issue until the site certificate information can be accessed by users?
Thank you!

**Comment by vesta0** [comment]
@softvision-miralobontiu this is good enough for now as we have Fennec parity. I will adjust the acceptance criteria. Thank you!

**Comment by feedback-query122** [comment]
@vesta0 desktop already has the ability to view certificates, and recently shipped a new in-content certificate viewer. However, the feature was never added to Firefox for Android while many other mobile browsers had the facility. Adding the feature to Fenix will be a milestone in feature parity and should not be cancelled. Closing this issue will only lead to delay or, perhaps, in non-implementation. I believe keeping this issue open is the best way for faster development and eventual implementation. Please reconsider opening it. Thank you!

**Comment by feedback-query122** [comment]
@vesta0 desktop already has the ability to view certificates, and recently shipped a new in-content certificate viewer. However, the feature was never added to Firefox for Android while many other mobile browsers had the facility. Adding the feature to Fenix will be a milestone in feature parity and should not be cancelled. Closing this issue will only lead to delay or, perhaps, non-implementation. I believe keeping this issue open is the best way for faster development and eventual implementation. Please reconsider opening it. Thank you!

**Comment by vesta0** [comment]
Thanks for this feedback! I have opened  #8400 to make sure this doesn't fall through the cracks.
